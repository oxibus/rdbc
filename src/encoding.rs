// https://github.com/hsivonen/recode_rs

use std::io::{Read, Write};
use std::str::from_utf8_mut;

use encoding_rs::*;

use crate::error::DbcError;

pub fn utf8_to_gbk(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "UTF-8", "GBK")
}

pub fn gbk_to_utf8(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "GBK", "UTF-8")
}

pub fn to_utf8(src_encoding_label: &str, src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, src_encoding_label, "UTF-8")
}

pub fn recode(
    src_data: &[u8],
    src_encoding_label: &str,
    dst_encoding_label: &str,
) -> Result<Vec<u8>, DbcError> {
    let src_encoding = get_encoding(Some(src_encoding_label.to_string()))?;
    let dst_encoding = get_encoding(Some(dst_encoding_label.to_string()))?;

    let mut decoder = src_encoding.new_decoder();
    let mut encoder = dst_encoding.new_encoder();

    let mut buf = std::io::Cursor::new(Vec::new());

    convert_via_utf8(
        &mut decoder,
        &mut encoder,
        &mut std::io::Cursor::new(src_data),
        &mut buf,
        true,
    )?;

    Ok(buf.into_inner())
}

pub fn get_encoding(opt: Option<String>) -> Result<&'static Encoding, DbcError> {
    match opt {
        None => Ok(UTF_8),
        Some(label) => match Encoding::for_label(label.as_bytes()) {
            None => Err(DbcError::InvalidEncodingLabel(label)),
            Some(encoding) => Ok(encoding),
        },
    }
}

pub fn convert_via_utf8(
    decoder: &mut Decoder,
    encoder: &mut Encoder,
    read: &mut dyn Read,
    write: &mut dyn Write,
    last: bool,
) -> Result<(), DbcError> {
    let mut input_buffer = [0u8; 2048];
    let mut intermediate_buffer_bytes = [0u8; 4096];
    let intermediate_buffer: &mut str = from_utf8_mut(&mut intermediate_buffer_bytes[..]).unwrap();
    let mut output_buffer = [0u8; 4096];
    let mut current_input_ended = false;
    while !current_input_ended {
        match read.read(&mut input_buffer) {
            Err(e) => {
                log::error!("Error reading input, error = {}", e);
                return Err(DbcError::EncodingReadInputError);
            }
            Ok(decoder_input_end) => {
                current_input_ended = decoder_input_end == 0;
                let input_ended = last && current_input_ended;
                let mut decoder_input_start = 0usize;
                loop {
                    let (decoder_result, decoder_read, decoder_written, _) = decoder.decode_to_str(
                        &input_buffer[decoder_input_start..decoder_input_end],
                        intermediate_buffer,
                        input_ended,
                    );
                    decoder_input_start += decoder_read;

                    let last_output = if input_ended {
                        match decoder_result {
                            CoderResult::InputEmpty => true,
                            CoderResult::OutputFull => false,
                        }
                    } else {
                        false
                    };

                    // Regardless of whether the intermediate buffer got full
                    // or the input buffer was exhausted, let's process what's
                    // in the intermediate buffer.

                    if encoder.encoding() == UTF_8 {
                        // If the target is UTF-8, optimize out the encoder.
                        if let Err(e) =
                            write.write_all(&intermediate_buffer.as_bytes()[..decoder_written])
                        {
                            log::error!("Error writing output, error = {}", e);
                            return Err(DbcError::EncodingWriteOutputError);
                        }
                    } else {
                        let mut encoder_input_start = 0usize;
                        loop {
                            let (encoder_result, encoder_read, encoder_written, _) = encoder
                                .encode_from_utf8(
                                    &intermediate_buffer[encoder_input_start..decoder_written],
                                    &mut output_buffer,
                                    last_output,
                                );
                            encoder_input_start += encoder_read;
                            if let Err(e) = write.write_all(&output_buffer[..encoder_written]) {
                                log::error!("Error writing output, error = {}", e);
                                return Err(DbcError::EncodingWriteOutputError);
                            }
                            match encoder_result {
                                CoderResult::InputEmpty => {
                                    break;
                                }
                                CoderResult::OutputFull => {
                                    continue;
                                }
                            }
                        }
                    }

                    // Now let's see if we should read again or process the
                    // rest of the current input buffer.
                    match decoder_result {
                        CoderResult::InputEmpty => {
                            break;
                        }
                        CoderResult::OutputFull => {
                            continue;
                        }
                    }
                }
            }
        }
    }

    Ok(())
}
