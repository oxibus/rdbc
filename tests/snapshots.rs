#![cfg(feature = "serde")]

use std::fs;
use std::path::Path;

use insta::{assert_debug_snapshot, assert_snapshot, assert_yaml_snapshot, with_settings};
use rrdbc::ast::network_ast::parse_dbc;
use rrdbc::encoding::decode_cp1252;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        ("dbc/dbc_from_cantools", "dbc-cantools"),
        ("dbc/mytest", "mytest"),
    ];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(&path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: &Path) {
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match parse_dbc(buffer.as_ref()) {
            Ok(dbc) => assert_yaml_snapshot!(file_name, dbc),
            Err(e) => {
                eprintln!("Failed to parse {file_name}.dbc: {e:?}");
                assert_debug_snapshot!(format!("!error___{file_name}"), e);
            }
        }
    } else {
        let error = format!("Failed to decode {file_name}.dbc as cp1252");
        eprintln!("{error}");
        assert_snapshot!(format!("!error___{file_name}"), error);
    }
}
