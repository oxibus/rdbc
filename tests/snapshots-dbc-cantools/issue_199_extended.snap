---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing:
  value: ~
nodes:
  - GMLAN
  - NEO
value_tables:
  - name: GearShifter
    value_descriptions:
      values:
        - num: 3
          str: Park
        - num: 0
          str: Drive/Low
  - name: DriverDoorStatus
    value_descriptions:
      values:
        - num: 1
          str: Opened
        - num: 0
          str: Closed
  - name: LKAGapButton
    value_descriptions:
      values:
        - num: 2
          str: "???"
        - num: 1
          str: "??"
        - num: 0
          str: None
  - name: CruiseButtons
    value_descriptions:
      values:
        - num: 6
          str: Cancel
        - num: 5
          str: Main
        - num: 3
          str: Set
        - num: 2
          str: Resume
        - num: 1
          str: None
  - name: CruiseControlActive
    value_descriptions:
      values:
        - num: 1
          str: Active
        - num: 0
          str: Inactive
  - name: BlinkerStatus
    value_descriptions:
      values:
        - num: 1
          str: Active
        - num: 0
          str: Inactive
messages:
  - header:
      id: 4294967295
      name: DriverDoorStatus
      size: 1
      transmitter: GMLAN
    signals:
      - name: DriverDoorOpened
        multiplexer: ~
        start_bit: 0
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
  - header:
      id: 272629760
      name: Chime
      size: 5
      transmitter: NEO
    signals:
      - name: ChimeType
        multiplexer: ~
        start_bit: 7
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - GMLAN
      - name: ChimeRepeat
        multiplexer: ~
        start_bit: 23
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - GMLAN
      - name: ChimeDuration
        multiplexer: ~
        start_bit: 15
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - GMLAN
      - name: ChimeByte5
        multiplexer: ~
        start_bit: 39
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - GMLAN
      - name: ChimeByte4
        multiplexer: ~
        start_bit: 31
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - GMLAN
  - header:
      id: 270581760
      name: BlinkerStatus
      size: 5
      transmitter: GMLAN
    signals:
      - name: RightBlinker
        multiplexer: ~
        start_bit: 6
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
      - name: LeftBlinker
        multiplexer: ~
        start_bit: 7
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
      - name: BlinkerLight
        multiplexer: ~
        start_bit: 25
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
  - header:
      id: 270794752
      name: SteeringWheelAngle
      size: 8
      transmitter: GMLAN
    signals:
      - name: SteeringWheelAngle
        multiplexer: ~
        start_bit: 39
        size: 16
        byte_order: BigEndian
        value_type: Signed
        factor: 0.0625
        offset: 0
        min: -540
        max: 540
        unit: deg
        receivers:
          - NEO
  - header:
      id: 271368192
      name: GearShifter
      size: 8
      transmitter: GMLAN
    signals:
      - name: GearShifter
        multiplexer: ~
        start_bit: 17
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - NEO
  - header:
      id: 271360000
      name: GasPedalRegenCruise
      size: 8
      transmitter: GMLAN
    signals:
      - name: CruiseControlActive
        multiplexer: ~
        start_bit: 56
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - GMLAN
      - name: MaxRegen
        multiplexer: ~
        start_bit: 12
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - GMLAN
          - NEO
      - name: GasPedal
        multiplexer: ~
        start_bit: 47
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 254
        unit: ""
        receivers:
          - GMLAN
          - NEO
      - name: GearShifter2NotUsed
        multiplexer: ~
        start_bit: 55
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - GMLAN
          - NEO
  - header:
      id: 270860288
      name: BrakePedal
      size: 2
      transmitter: GMLAN
    signals:
      - name: BrakeLevel
        multiplexer: ~
        start_bit: 2
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - NEO
      - name: BrakeSensor
        multiplexer: ~
        start_bit: 15
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - NEO
  - header:
      id: 275480576
      name: WheelSpeed
      size: 8
      transmitter: GMLAN
    signals:
      - name: WheelSpeedFL
        multiplexer: ~
        start_bit: 7
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 70
        unit: yd/s
        receivers:
          - NEO
      - name: WheelSpeedFR
        multiplexer: ~
        start_bit: 39
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 70
        unit: yd/s
        receivers:
          - NEO
      - name: WheelSpeedRL
        multiplexer: ~
        start_bit: 23
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 70
        unit: yd/s
        receivers:
          - NEO
      - name: WheelSpeedRR
        multiplexer: ~
        start_bit: 55
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 70
        unit: yd/s
        receivers:
          - NEO
  - header:
      id: 270598144
      name: VehicleSpeed
      size: 8
      transmitter: GMLAN
    signals:
      - name: VehicleSpeed1
        multiplexer: ~
        start_bit: 7
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 100
        unit: mph
        receivers:
          - NEO
      - name: VehicleSpeed2
        multiplexer: ~
        start_bit: 39
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 100
        unit: mph
        receivers:
          - NEO
  - header:
      id: 276135936
      name: CruiseButtons
      size: 3
      transmitter: GMLAN
    signals:
      - name: CruiseButtons
        multiplexer: ~
        start_bit: 3
        size: 3
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 12
        unit: ""
        receivers:
          - NEO
  - header:
      id: 276127744
      name: CruiseButtons2
      size: 1
      transmitter: GMLAN
    signals:
      - name: LKAGapButton
        multiplexer: ~
        start_bit: 1
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 2
        unit: ""
        receivers:
          - NEO
env_vars: []
env_vars_data: []
comments: []
attribute_definitions:
  - Network:
      attribute_name: UseGMParameterIDs
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 0
  - Network:
      attribute_name: ProtocolType
      attribute_value_type:
        String: {}
  - Network:
      attribute_name: BusType
      attribute_value_type:
        String: {}
attribute_defaults:
  - Attribute:
      attribute_name: UseGMParameterIDs
      attribute_value:
        Double: 1
  - Attribute:
      attribute_name: ProtocolType
      attribute_value:
        String: GMLAN
  - Attribute:
      attribute_name: BusType
      attribute_value:
        String: ""
attribute_values:
  - Network:
      attribute_name: BusType
      attribute_value:
        String: CAN
  - Network:
      attribute_name: ProtocolType
      attribute_value:
        String: GMLAN
signal_value_descriptions:
  - message_id: 4294967295
    signal_name: DriverDoorOpened
    value_descriptions:
      values:
        - num: 1
          str: Opened
        - num: 0
          str: Closed
  - message_id: 270581760
    signal_name: RightBlinker
    value_descriptions:
      values:
        - num: 1
          str: Active
        - num: 0
          str: Inactive
  - message_id: 270581760
    signal_name: LeftBlinker
    value_descriptions:
      values:
        - num: 1
          str: Active
        - num: 0
          str: Inactive
  - message_id: 270581760
    signal_name: BlinkerLight
    value_descriptions:
      values:
        - num: 1
          str: Active
        - num: 0
          str: Inactive
  - message_id: 271368192
    signal_name: GearShifter
    value_descriptions:
      values:
        - num: 3
          str: Park
        - num: 0
          str: Drive/Low
  - message_id: 271360000
    signal_name: CruiseControlActive
    value_descriptions:
      values:
        - num: 1
          str: Active
        - num: 0
          str: Inactive
  - message_id: 276135936
    signal_name: CruiseButtons
    value_descriptions:
      values:
        - num: 6
          str: Cancel
        - num: 5
          str: Main
        - num: 3
          str: Set
        - num: 2
          str: Resume
        - num: 1
          str: None
  - message_id: 276127744
    signal_name: LKAGapButton
    value_descriptions:
      values:
        - num: 2
          str: "???"
        - num: 1
          str: "??"
        - num: 0
          str: None
env_var_value_descriptions: []
